scr/src/main/java/org/apache/felix/scr/impl/manager/ComponentContextImpl.java

```diff
@@ -328,7 +328,7 @@ public void dispose()
 
     private Map<RefPair<?, ?>, Object> createNewFieldHandlerMap()
     {
-        return new TreeMap<>(
+        return new ConcurrentSkipListMap<>(
```

Comment:
While I think this will avoid the exception in TreeMap, I do not think it is a complete solution.  In `org.apache.felix.scr.impl.inject.field.FieldHandler.updateField(METHOD_TYPE, Object, BindParameters)` we have the following code:

```java
            if ( mType == METHOD_TYPE.UNBIND )
            {
                if ( this.metadata.isOptional() && !this.metadata.isStatic() )
                {
                    // we only reset if it was previously set with this value
                    if ( bp.getComponentContext().getBoundValues(metadata.getName()).size() == 1 )
                    {
                        this.setFieldValue(componentInstance, null);
                    }
                }
                bp.getComponentContext().getBoundValues(metadata.getName()).remove(refPair);
            }
```
`getBoundValues` call here will return this `ConcurrentSkipListMap`.  If multiple threads are entering this block then I think we have an issue with never calling `this.setFieldValue(componentInstance, null);` when two threads enter the `size() ==1` check and find the size to be 2.  Both threads will remove their own refPair from the Map but none will null out the field value when the last is removed.

Reply:
Isn't the method that calls createNewFieldHandlerMap() synchronized?

Reply:
Ignore the above comment :) But I think in general the code expects to not be called concurrently (if I remember correctly). So there should be some synchronization somewhere

Reply:
@tjwatson According to the current implementation, `org.apache.felix.scr.impl.inject.field.FieldHandler.updateField(METHOD_TYPE, Object, BindParameters)` will be invoked by different implementations of `org.apache.felix.scr.impl.inject.field.FieldHandler.State.invoke(FieldHandler, METHOD_TYPE, Object, BindParameters)` and if we traverse back, none of the consumers of the aforementioned `invoke(..)` method is thread-safe (synchronized). I agree with you, we should introduce thread-safety in `org.apache.felix.scr.impl.inject.field.FieldHandler.updateField(METHOD_TYPE, Object, BindParameters)`.

Reply:
Added a new commit to make the `org.apache.felix.scr.impl.inject.field.FieldHandler.updateField(METHOD_TYPE, Object, BindParameters)` synchronized.